rslib/src/scheduler/fsrs/weights.rs

```diff
@@ -110,16 +123,23 @@ impl Collection {
         Ok(())
     }
 
-    pub fn evaluate_weights(&mut self, weights: &Weights, search: &str) -> Result<ModelEvaluation> {
+    pub fn evaluate_weights(
+        &mut self,
+        weights: &Weights,
+        search: &str,
+        ignore_revlogs_before_ms: TimestampMillis,
+    ) -> Result<ModelEvaluation> {
         let timing = self.timing_today()?;
         let mut anki_progress = self.new_progress_handler::<ComputeWeightsProgress>();
         let guard = self.search_cards_into_table(search, SortMode::NoOrder)?;
-        let revlogs = guard
+        let revlogs: Vec<RevlogEntry> = guard
             .col
             .storage
             .get_revlog_entries_for_searched_cards_in_card_order()?;
-        anki_progress.state.fsrs_items = revlogs.len() as u32;
-        let items = fsrs_items_for_training(revlogs, timing.next_day_at);
+        let filtered_revlogs = remove_revlogs_before(revlogs, ignore_revlogs_before_ms);
```

Comment:
I am not sure if the revlogs should also be filtered when evaluating the weights.
- In the Python version, if the card was first reviewed before the revlog_start_date, then all of its reviews are filtered out (because of the code that filters out incomplete revlogs).
- In the Rust version, cards with incomplete revlogs are treated better because their evaluation corresponds with how the cards will actually be scheduled.
But due to this difference, after the proposed change, the cards which were first reviewed before the "start date" will have their memory states calculated from SM-2 variables during evaluation and using normal formulas during scheduling.

Reply:
>In the Python version, if the card was first reviewed before the revlog_start_date, then all of its reviews are filtered out (because of the code that filters out incomplete revlogs).

Python code in question:
https://github.com/open-spaced-repetition/fsrs-optimizer/blob/951981130b53dd023f6986127e7898b815e5c0d5/src/fsrs_optimizer/fsrs_optimizer.py#L663-L666        
```py
df = df.groupby("card_id").filter(
            lambda group: group["review_time"].min()
            > time.mktime(datetime.strptime(revlog_start_date, "%Y-%m-%d").timetuple())
            * 1000
        )
```

That does seem to be the case. That would explain why my deck optimized using the external and internal optimizer still differed when using this even with the same ignore date (Something I chalked up to differences in the rust and python implementations). As I see it my way is superior because otherwise cards that are reviewed before the ignore before date aren't completely ignored by the optimizer which ruins the idea of giving the card a "second chance". This would limit the cards the optimizer uses to the ones only added after the date. 

> In the Rust version, cards with incomplete revlogs are treated better than the Python version because their evaluation corresponds with how the cards will actually be scheduled.
But this strength of the Rust version also introduces new complications when making changes. After the proposed change, the cards which were first reviewed before the "start date" will have their memory states calculated from SM-2 variables during evaluation and using normal formulas during scheduling.

This change does not effect scheduling, it merely effects the "Evaluate" button on the deck config menu. I don't believe The Evaluate button is affected by the card states if that is what you're saying?

Reply:
> As I see it my way is superior because otherwise cards that are reviewed before the ignore before date aren't completely ignored by the optimizer which ruins the idea of giving the card a "second chance". 

The Rust optimizer also ignores such cards. The difference is in Evaluation.

> This change does not effect scheduling, it merely effects the "Evaluate" button on the deck config menu.

Yes, your change doesn't affect the scheduling. Before and after your change, the cards that were introduced before the start date will have their memory states calculated using normal FSRS formulas (if they have complete revlogs). 

However, after your change, during the evaluation, the memory states of such cards are calculated using SM-2 variables (which the scheduler uses when it encounters incomplete revlogs).

> I don't believe The Evaluate button is affected by the card's memory states if that is what you're saying?

It is not affected by the current memory states. But, it is affected by the memory states generated at each step of the simulation of the review history that happens when the user clicks `Evaluate`.

Reply:
I made it so that it adds a fake learning review to the start of every truncated review which should prevent sm2 being used and it seems to have made no difference.

```rs
// weights.rs:29 
fn remove_revlogs_before(revlogs: Vec<RevlogEntry>, ms: TimestampMillis) -> Vec<RevlogEntry> {

    // If there is no minimum time or the minimum time is lower than every review's review time.
    if ms.0 == 0 || revlogs.iter().map(|revlog| revlog.id.0).min().unwrap_or(-1) > ms.0 {
        revlogs
    } else {
        // Insert a fake revlog at the start so single_card_revlog_to_items doesn't use sm2.
        let filtered_revlogs = revlogs.into_iter().filter(|revlog| revlog.id.0 > ms.0);

        let mut out = vec![RevlogEntry {
            id: ms.into(),
            cid: 0.into(),
            usn: 0.into(),
            button_chosen: 1,
            interval: -1,
            last_interval: -1,
            ease_factor: 250,
            taken_millis: 0,
            review_kind: RevlogReviewKind::Learning,
        }];

        out.extend(filtered_revlogs);
        out
    }
}
```

With change:
![image](https://github.com/ankitects/anki/assets/63685643/242f10b4-ea41-41cb-9eac-50141f562b0b)

Without change:
![image](https://github.com/ankitects/anki/assets/63685643/0a5230bb-c19b-49a6-9ef6-0c2cdcee2953)

Unless I misunderstood the detection code or something else:

https://github.com/ankitects/anki/blob/3982e0c8fe691d3e84deceb7555f2d2fd0a8fbe6/rslib/src/scheduler/fsrs/weights.rs#L177-L188

I don't think this affects the reviews at all.
This was still worth looking into though. Thanks.

Reply:
@L-M-Sherlock @expertium can you also confirm you're happy with the approach this PR takes?

Reply:
> That does seem to be the case. That would explain why my deck optimized using the external and internal optimizer still differed when using this even with the same ignore date (Something I chalked up to differences in the rust and python implementations). As I see it my way is superior because otherwise cards that are reviewed before the ignore before date aren't completely ignored by the optimizer which ruins the idea of giving the card a "second chance". This would limit the cards the optimizer uses to the ones only added after the date.

I don't agree. What if the user use `hard` as `again` before the ignore date? These reviews will be used as history features in your implementation. They will pollute the dataset and make FSRS confused.

Reply:
I'll change it to the python method if you want but I'm still confused about the mechanics of this.
In a revlog where | represents the ignore before date:

```
Original revlog:              1 3 3 2 2 2 | 1 1 3 3 1 3
Current implementation:                   | 1 1 3 3 1 3
```
I want to know exactly where the pollution comes from here. Did I miss some place where the SM-2 values are substituted in or something? What would I need to change to remove this pollution?
Using the python implementation this entire revlog would be ignored which means this feature would be rendered useless for people who didn't add any cards after the ignore before because it would ignore every card, right?

Reply:
Sorry, I misunderstood the current implementation. The current problem is the incomplete revlogs. In FSRS, the first rating is treated in different way than subsequent ratings. The current implementation will let FSRS use review data in pretraining process, which is incorrect.

Reply:
In my opinion, the best solution is to make the evaluation exactly match the scheduler behaviour. This is because the main purpose of the evaluation is to determine the effectiveness of scheduling, which would be most accurate when both treat the revlogs in the same way.

To do this, the revlogs before the start date should NOT be removed during evaluation.

> Using the python implementation this entire revlog would be ignored which means this feature would be rendered useless for people who didn't add any cards after the ignore before because it would ignore every card, right?

The optimizer will still completely ignore those cards, which means that the issue will persist. However, incomplete revlogs are not usable for training. Refer to https://github.com/open-spaced-repetition/fsrs-rs/issues/114 for a more detailed description.

Reply:
> The current implementation will let FSRS use review data in pretraining process, which is incorrect.

@L-M-Sherlock, I don't think so because the first entry in the splitted revlog won't be of LEARN type. However, I may be interpreting the code incorrectly. So, you can double-check.

Reply:
> In FSRS, the first rating is treated in different way than subsequent ratings

So while filtering the revlogs, The first revlog should be preserved even if its before the date?

> I hope that this was for testing purpose only and you don't want to include it in the PR.

Yep just testing.

> In my opinion, the best solution is to make the evaluation exactly match the scheduler behavior. This is because the main purpose of the evaluation is to determine the effectiveness of scheduling, which would be most accurate when both treat the revlogs in the same way.

Would a better alternative solution be to make it so that the scheduler ignores these cards too therefore having the scheduler treat both the same way? I apparently forgot that revlogs were used during scheduling not just optimization earlier 😖.

Reply:
> Would a better alternative solution be to make it so that the scheduler ignores these cards too 

I don't think that there is a major downside to this approach except that we will have to use memory_state_from_sm2 (which is not very reliable) in a larger number of situations. But, if the past review history is very different from the current one, this can be a reasonable compromise (which the user has to make by choosing a revlog start date).

However, I still want to know the opinion of @L-M-Sherlock on this one.

Reply:
Alright, I programmed it to affect scheduling too. If this is a bad change I'll revert it and just remove the evaluation filter code.

Reply:
> I don't think so because the first entry in the splitted revlog won't be of LEARN type.

FSRS itself doesn't have any filters except the outlier filter. Other filters are implemented in Anki.